### PR TITLE
feat(hc): Parallel RPCs to multiple regions (spike)

### PR DESCRIPTION
**Pull request by RyanSkonnord**
This was a brief investigation into how much code complexity would be involved in having an RPC that automagically "fans out" from the control silo to multiple regions.

There are several reasons not to actually do this. Performance isn't necessarily one of them: we can mitigate the time costs of multiple RPCs by making them in parallel, and the overall cost in system load isn't necessarily a concern. But having each call to a Hybrid Cloud service cause at most one RPC is valuable as a simplifying principle. Multiple calls make failure modes more complex: if we query several regions and one fails, does the entire operation fail? When would it make sense to degrade gracefully with a partial set of results? (Thanks @corps for pointing this out.) Better to force the callers to figure out both whether it's truly worth calling multiple regions and how to compose the results if so.

But I wanted to do the spike anyway, so please review if you're curious. It's an interesting option to keep in our back pocket and we may get the chance to reuse some of the code for parallel RPCs in a narrower, more deliberately chosen context.

### CODE REVIEW

File: src/sentry/services/hybrid_cloud/rpc.py [hunk 1]
```diff
@@ -5,7 +5,23 @@
 import urllib.response
 from abc import abstractmethod
 from dataclasses import dataclass
-from typing import TYPE_CHECKING, Any, Callable, Dict, Iterator, Mapping, Tuple, Type, TypeVar, cast
+from queue import Queue
+from threading import Thread
```
Comment (RyanSkonnord) [inline src/sentry/services/hybrid_cloud/rpc.py:9]: From the [Python docs](https://docs.python.org/3/library/threading.html):

> If you want your application to make better use of the computational resources of multi-core machines, you are advised to use [`multiprocessing`](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing) or [`concurrent.futures.ProcessPoolExecutor`](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ProcessPoolExecutor). However, `threading` is still an appropriate model if you want to run multiple I/O-bound tasks simultaneously.

**Comment by mikejihbe** [comment]
This is an anti-pattern. We should try hard _not_ to do this. Any use cases that require it would most likely be better refactored to avoid this. I would almost prefer not to have the tool available unless we have a concrete use for it.

**Comment by RyanSkonnord** [comment]
> This is an anti-pattern. We should try hard _not_ to do this. Any use cases that require it would most likely be better refactored to avoid this. I would almost prefer not to have the tool available unless we have a concrete use for it.

In case I didn't make it clear enough in the description, I agree.

This came up for me in the context of `OrganizationService.get_organizations` (note the [problem](https://github.com/getsentry/sentry/blob/a149a91/src/sentry/services/hybrid_cloud/organization/service.py#L64) with region resolution) and the `User` model's `get_orgs` and `get_projects` methods. If you like, you can read this PR as an essay on, "if we *don't* want to tear up and change those methods' individual call sites, this is what we would have to do instead."

File: src/sentry/services/hybrid_cloud/organization/service.py [hunk 2]
```diff
@@ -16,10 +18,30 @@
...
+
+class ByOrganizationMembership(RegionResolution):
+    def resolve(self, arguments: ArgumentDict) -> Collection[Region]:
+        from sentry.models import OrganizationMemberMapping
+
+        user_id: int = arguments["user_id"]
+        member_mappings = OrganizationMemberMapping.objects.filter(user_id=user_id)
+        organization_ids = member_mappings.values_list("organization_id", flat=True)
+        mappings = self.organization_mapping_manager.filter(organization_id__in=organization_ids)
+        return [self._resolve_from_mapping(mapping) for mapping in mappings]
```
Comment (dashed) [inline src/sentry/services/hybrid_cloud/organization/service.py:37]: Just curious, do we do de-duplication on the list of regions?

Comment (RyanSkonnord) [reply src/sentry/services/hybrid_cloud/organization/service.py:37]: Oh right, good point, it should do that.

**Pull request closed without merging**